e2e/main.py

```diff
@@ -79,6 +88,126 @@ def auction_avl_leaves(
             yield ptr, leaf_node
 
 
+# A Profiler is a helper function for wrapping a call to a contract with logic
+# for extracting gas cost and other related information.
+Profiler: Callable[[PyTezosClient, ContractInterface, List[OperationGroup]], Dict]
+
+
+def general_gas_profiler(
+    client: PyTezosClient, checker: ContractInterface, op_batch: List[OperationGroup]
+) -> Dict:
+    """Calls inject for a list of entrypoint calls, recording the gas cost of each call."""
+    ret = inject(
+        client,
+        client.bulk(*op_batch).autofill(ttl=MAX_OPERATIONS_TTL).sign(),
+    )
+    profile = {}
+    for op in ret["contents"]:
+        name = op["parameters"]["entrypoint"]
+        if name not in profile:
+            profile[name] = {"gas": []}
+        profile[name]["gas"].append(op["gas_limit"])
+    return profile
+
+
+def mark_for_liquidation_profiler(
+    client: PyTezosClient, checker: ContractInterface, op_batch: List[OperationGroup]
+) -> Dict:
+    """
+    Calls inject for a list of `mark_for_liquidation` calls, recording
+    the gas cost and current liquidation queue size for each call.
+    """
+    n = len(
+        list(
+            auction_avl_leaves(
+                checker,
+                checker.storage["deployment_state"]["sealed"]["liquidation_auctions"][
+                    "queued_slices"
+                ](),
+            )
+        )
+    )
+    ret = inject(
+        client,
+        client.bulk(*op_batch).autofill(ttl=MAX_OPERATIONS_TTL).sign(),
+    )
+
+    profile = {
+        "mark_for_liquidation": {
+            "gas": [],
+            "queue_size": [],
+        }
+    }
+    for op in ret["contents"]:
+        assert op["parameters"]["entrypoint"] == "mark_for_liquidation"
+        profile["mark_for_liquidation"]["gas"].append(op["gas_limit"])
+        profile["mark_for_liquidation"]["queue_size"].append(n)
+        # Assuming that each request_liquidation operation adds a single slice to the queue
+        # to avoid having to re-query the storage each time (which can take a long time).
```

Comment:
Oh that is a very good point that I hadn't thought of. As you note, I do think it at least holds in the one place where it's called. I'm not sure if there is an efficient way to assert that this is always correct though since querying storage can be slow.

Reply:
Yeah, I understand. For now it should be OK to just adjust the comment to mention that this operation does not always add a single slice; if we ever need a more general-purpose functionality we can revisit this ;)